### PR TITLE
update: Refactor client-side Layout code and add some common type definitions

### PR DESCRIPTION
**Pull request by idreyn**
This is probably work on #1052 but I decided to cut off a PR here before it gets too out of hand. Basically, I've moved the meat of the `Page` component into a `Layout` component that can also be consumed by a soon-to-emerge Collection container component, and generalized the `PubsBlock` component into a `BasePubsBlock` which will also be consumed by the new block type, `CollectionBlock`. In service of all of this, I created a new `utils/types.ts` file where we will host common types related to our data model, applied it in a few places, and cleaned up obsolete `@ts-expect-error` comments.

_Test plan:_
Not much here since this is pretty intermediate — if we can get a Page and a Page editor to render, that's good enough for now.

### CODE REVIEW

File: client/components/Layout/BaseLayoutPubs.tsx [hunk 1]
```diff
@@ -0,0 +1,87 @@
+import React from 'react';
```
Comment (idreyn) [inline client/components/Layout/BaseLayoutPubs.tsx:1]: Most of this component is taken from `LayoutPubs.tsx`

File: client/components/Layout/Layout.tsx [hunk 2]
```diff
@@ -0,0 +1,77 @@
+import React from 'react';
```
Comment (idreyn) [inline client/components/Layout/Layout.tsx:1]: Most of this is taken from `Page.tsx`

File: utils/layout/util.js [hunk 3]
```diff
@@ -0,0 +1,120 @@
+export const validBlockTypes = [
+	'pubs',
+	'text',
+	'html',
+	'banner',
+	'pages', // TODO(ian): Remove this after migration
+	'collections-pages',
+];
+
+export const getPubsByBlockIndex = (blocks, pubs) => {
```
Comment (idreyn) [inline utils/layout/util.js:10]: This is `generateRenderLists` by a more descriptive name and moved from `utils/pages.js`

File: client/components/PubPreview/PubPreview.tsx [hunk 4]
```diff
@@ -2,6 +2,7 @@ import React, { useState, useRef, useEffect } from 'react';
 import classNames from 'classnames';
 import dateFormat from 'dateformat';
 
+import { Pub } from 'utils/types';
```
Comment (3mcd) [inline client/components/PubPreview/PubPreview.tsx:5]: 👌

File: utils/types.ts [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+	collectionPubs?: CollectionPub[];
+	exports?: Export[];
+	members?: Member[];
+	branches?: Branch[];
+	releases?: Release[];
+	pubVersions?: PubVersion[];
+	crossrefDepositRecord?: CrossrefDepositRecord;
+	inboundEdges?: InboundEdge[];
+	outboundEdges?: OutboundEdge[];
+};
```
Comment (3mcd) [inline utils/types.ts:175]: Love it!

**Review by 3mcd — APPROVED** [review]
LGTM! Awesome work on all of the types!

**Pull request merged**